pkg/arvo/neo/cod/std/src/fil/messenger.md

```diff
@@ -562,3 +563,273 @@ The `stem` is head-tagged with a `care:neo`. If the shrub’s dependent is liste
+
+```
+We’ll take a look at Messenger. Messenger is one shrub, located at `/imp/messenger.hoon`, but it relies on five other shrubs to work.
+- `/imp/messenger` is the high-level interface and “service provider” that enables the user to create groupchats and 1-to-1 DMs, and invite other ships to them.
+- `/imp/message` is a `~` stub that allows us to create `/message`s in the namespace
+- `/imp/message-pub` is a shrub that takes a poke to store `/message`s in the namespace as its kids.
+- `/imp/message-sub` mirrors `/message`s from a `/message-pub` shrub into its own state.
+- `/imp/groupchat` creates a publisher/subscriber pair, and can invite other ships to post/subscribe to the `/message-pub` shrub.
+- `/imp/dm` negotiates a two-way pub/sub relationship and mirrors state between both parties.
+```
+
+### DMs in shrubbery
+If you wanted to implement 1-on-1 DMs in your shrub, you could just include `/imp/dm` in your `+deps` arm and send `%make` and `%poke` cards to it. If that doesn't do what you need, you could base your own DM functionality on this.
+
+`/imp/dm` is short and simple enough that it might be nice to show an equivalent `/app/dm` Gall agent, but there isn't one. `/imp/dm` delegates a lot of functionality to service providers in a way that Gall apps theoretically could, but in practice never do. With the constraints they can set in the `+kids` and `+deps` arms, shrubs can make guarantees about the functionality of third-party service providers in other shrubs and/or other desks. Those service providers can make breaking changes to that functionality, but as long as they have some way of converting data to the type that their parents or dependents accept, those shrubs will still be able to interoperate.
+
+#### kook:neo
+
+```hoon
+::  /imp/dm.hoon
+++  state  [%pro %ship]
+++  poke  (sy %dm-diff ~)
+++  kids
+  %-  some
+  :-  %y
+  %-  ~(gas by *lads:neo)
+  :~  :-  [[%.n %theirs] %.n]
+      [[%pro %message-pub] (sy %sig ~)]
+      :-  [[%.n %mine] %.n]
+      [[%pro %message-sub] (sy %sig ~)]
+  ==
+++  deps  *deps:neo
+++  form
+  ...
+```
+
+A DM shrub only stores one ship, the `@p` of whoever you're chatting with. It only has two kids: `/path/to/this/dm/theirs` and `/path/to/this/dm/mine`. It chooses the service providers at `/imp/message-pub` and `/imp/message-sub` to store the state of `/theirs` and `/mine`, while setting no constraints on the types of poke those service providers can take. In other words, it delegates the functionality to two new instances of `/imp/message-pub` and `/imp/message-sub` rather than subscribing to existing Gall agents that may or may not exist.
+
+```hoon
+$%  [%initiate partner=ship provider=pith]
+    [%invited partner=ship dm=pith]
+  ::
+    [%acked dm=pith]
+  ::
+    [%post text=@t]
+==
+```
+
+`/imp/dm` takes four pokes: `%initiate`, `%invited`, `%acked`, and `%post`.
+- `%initiate`: we initiate a DM, specifying the other ship and the service provider we want to use.
```

Comment:
This would be a good place to define "service provider," stressing that "service provider" is a conception/design pattern specifically for this set of shrubs that may or may not generalize to other shrubs depending on how well it works and what true service discovery looks like

Reply:
Specify that we could do this manually via making from the dojo, but in practice here, our parent shrub `messenger` will initialize this state in its own opinionated way